articles/realtime.md

```diff
@@ -0,0 +1,259 @@
+
+### Allocate dynamic memory pool
+
+```c
+  if (mlockall(MCL_CURRENT | MCL_FUTURE))
+    perror("mlockall failed:");
+
+  /* Turn off malloc trimming.*/
+  mallopt(M_TRIM_THRESHOLD, -1);
+
+  /* Turn off mmap usage. */
+  mallopt(M_MMAP_MAX, 0);
+
+  page_size = sysconf(_SC_PAGESIZE);
+  buffer = malloc(SOMESIZE);
+
+  for (i=0; i < SOMESIZE; i+=page_size)
+  {
+      buffer[i] = 0;
+  }
+  free(buffer);
+```
+
+It is commonly believed that dynamic memory allocations are not permitted in the real-time code path.
+This code snippet shows how to lock the virtual address space, disallow returning deallocated memory to the kernel via `sbrk`, and disable `mmap`.
+It effectively locks a pool of memory in the heap into RAM, which prevents page faults due to `malloc` and `free`.
+
+Pros:
+
+* Can use malloc/new, free/delete, and even STL containers
+
+Cons:
+
+* Must accurately predict bounded memory size for the process!
+* Using STL containers is therefore dangerous (unbounded sizes)
+* In practice, only works for small processes
+
+### Custom fixed allocators for STL containers
+
+An alternative to the above approach is to implement custom allocators for STL containers that only allocate memory on the stack.
+
+There are various comprehensive tutorials [already written](http://www.codeguru.com/cpp/article.php/c18503/C-Programming-Stack-Allocators-for-STL-Containers.htm) for this task.
+
+Pros:
+* Use existing STL code with deterministic computation and without allocating global dynamic memory pool
+* More modular solution
+
+Cons:
+* Complex to implement
+
+### Global variables and (static) arrays
+
+Global variables are preallocated at the start of a process, thus assigning and accessing them is real-time safe.
+However, this strategy comes with the many disadvantages of using global variables.
+
+### Use inheritance sparingly
+
+Classes with many levels of inheritance may not be real-time safe because of vtable overhead access.
+When executing an inherited function, the program needs to access the data used in the function, the vtable for the class, and the instructions for the function, which are all stored in different parts of memory, and may or may not be stored in cache together.
+
+In general, C++ patterns with poor cache locality are not well-suited to real-time environments.
+Another such pattern is the opaque pointer idiom (PIMPL), which is convenient for ABI compatibility and speeding up compile times.
+However, bouncing between the memory location for the object and its private data pointer causes the cache to "spill" as it loads one chunk of memory and then another, unrelated chunk for almost every function in the PIMPLized object.
+
+### Exceptions
+
+Throwing an exception can put large objects on the stack, which is undesirable in real-time programming since we cannot allocate memory on the heap.
```

Comment:
I don't really get this statement (stack vs. heap). My understanding is that when no exceptions are thrown, the overhead is small, but exception handling (stack unwinding, catching) can add a significant penalty. FWIW, this [stack overflow](http://stackoverflow.com/a/2226262) answer states something similar.